src/nodeViews/createReactNodeViewConstructor.tsx

```diff
@@ -130,6 +158,16 @@ export function createReactNodeViewConstructor(
         initialState.isSelected
       );
 
+      const portalRegistry = useContext(PortalRegistryContext);
+      const childRegisteredPortals = portalRegistry[key];
+      const childPortals = useMemo(
+        () =>
+          childRegisteredPortals
+            ?.sort((a, b) => a.getPos() - b.getPos())
+            .map(({ portal }) => portal),
+        [childRegisteredPortals]
+      );
```

Comment:
Can we maintain them in sorted order in the registry to make this easier?

Reply:
I _think_ the answer is no; we want to sort them when we render, not when we register, because the positions could change, right? Like if you swapped two nodes, ProseMirror might keep the node views around, so we wouldn't re-register, but the parent node view would get updated, so it would have a chance to check the new positions and put them in the right order

Reply:
I'm forgetting that the registrations here are for _descendants_, not just children, in the event that there are non-React node views between here and those descendants.

Is there something we could do so that there's a React component for every node view, even if it's a non-React node view, and then the React component tree would reflect the document easily? In `NodeViews` could we iterate over the children and render either their registered portals or a stub component that just does the same for its children? That might allow node views to not even register with an ancestor, but just register a portal for their key.

Reply:
This _seems_ like it should be possible, but I'm finding it challenging in practice. I can construct the initial tree (which is neat), but:

I started out trying to simply use `useEditorState` to retrieve the positionRegistry and look up the children of the current node:

```tsx
function useChildNodeViews(node: Node, pos: number) {
  const portals = useContext(PortalRegistryContext);

  const editorState = useEditorState();
  if (!editorState) return null;

  const positionRegistry = reactNodeViewPlugin.getState(editorState)!;
  console.log(positionRegistry);

  const childPositions: { node: Node; pos: number }[] = [];
  node.forEach((child, offset) => {
    childPositions.push({ node: child, pos: pos + 1 + offset });
  });

  return childPositions.map(({ node, pos }) => {
    const key = positionRegistry.get(pos)!;
    const portal = portals[key];
    return portal ?? <PlainNodeView key={key} node={node} pos={pos} />;
  });
}
```

This has a gnarly state tearing issue: the `getPos`  function passed to node views is based state in the view descriptors, so it's "stale" at render time compared to the plugin state returned by `useEditorState`. The result is that you get into infinite loops whenever you insert a character before an existing node; the node view keeps finding its _own portal_ in the registry, because it's using it's old position to look up its first child.

This led to a second approach that feels bad and doesn't work, but I think I'm too tired to figure out why.

```tsx
function useChildNodeViews(node: Node, getPos: () => number) {
  const portals = useContext(PortalRegistryContext);
  const [children, setChildren] = useState<JSX.Element[]>([]);

  useEditorEffect(
    (view) => {
      if (!view) return;
      console.log("in usechildnodes");

      const positionRegistry = reactNodeViewPlugin.getState(view.state)!;

      const childPositions: { node: Node; pos: number }[] = [];
      node.forEach((child, offset) => {
        childPositions.push({ node: child, pos: getPos() + 1 + offset });
      });

      setChildren(
        childPositions
          .filter(({ node }) => !node.isText)
          .map(({ node, pos }) => {
            const key = positionRegistry.get(pos)!;
            // console.log(node.type.name, node.textContent, key);
            const portal = portals[key];
            return portal ?? <PlainNodeView key={key} node={node} pos={pos} />;
          })
      );
    },
    [node, portals, getPos]
  );

  return children;
}
```

This way, the position isn't looked up until the state and view have been reconciled. This is... better? But it still isn't right. Some categories of changes to the doc (e.g. trying to split a list item when there's another list item after it) are totally broken, though basic editing more or less works.

I'll push this attempt to another branch so you can take a look if you like!

Reply:
https://github.com/nytimes/react-prosemirror/pull/28

Reply:
Thanks!